File: backend/app/schemas/imports.py
from datetime import date

from pydantic import BaseModel, field_validator

from app.models.observation import ObservationType


class ImportPreviewRequest(BaseModel):
    text: str


class ImportSummary(BaseModel):
    total_rows: int = 0
    importable: int = 0
    conflicts: int = 0
    errors: int = 0
    skipped: int = 0
    imported: int = 0


class ImportItem(BaseModel):
    row: int
    date: str
    type: ObservationType
    label: str
    incoming_value: str
    existing_value: str | None = None
    status: str
    conflict: bool = False
    error: str | None = None
    overwrite: bool = False

    @field_validator("date")
    @classmethod
    def validate_date(cls, value: str) -> str:
        date.fromisoformat(value)
        return value


class ImportApplyRequest(BaseModel):
    items: list[ImportItem]


class ImportRow(BaseModel):
    row: int
    date: str | None
    status: str
    message: str | None = None


class ImportPreviewResponse(BaseModel):
    rows: list[ImportRow]
    items: list[ImportItem]
    summary: ImportSummary


class ImportApplyResponse(BaseModel):
    items: list[ImportItem]
    summary: ImportSummary

File: frontend/src/api/imports.ts
import { handleUnauthorized } from './auth'

export type ImportSummary = {
  total_rows: number
  importable: number
  conflicts: number
  errors: number
  skipped: number
  imported: number
}

export type ImportItem = {
  row: number
  date: string
  type: string
  label: string
  incoming_value: string
  existing_value: string | null
  status: 'ready' | 'conflict' | 'error' | 'skipped' | 'imported'
  error: string | null
  overwrite: boolean
  conflict: boolean
}

export type ImportRow = {
  row: number
  date: string | null
  status: string
  message: string | null
}

export type ImportPreviewResponse = {
  rows: ImportRow[]
  items: ImportItem[]
  summary: ImportSummary
}

export type ImportApplyResponse = {
  items: ImportItem[]
  summary: ImportSummary
}

async function parseJson<T>(response: Response): Promise<T> {
  if (!response.ok) {
    if (response.status === 401) handleUnauthorized()
    throw new Error(String(response.status))
  }
  return response.json() as Promise<T>
}

export async function previewReadingsImport(accessToken: string, text: string): Promise<ImportPreviewResponse> {
  const response = await fetch('/api/import/observations/preview', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
    },
    credentials: 'include',
    body: JSON.stringify({ text }),
  })
  return parseJson<ImportPreviewResponse>(response)
}

export async function applyReadingsImport(accessToken: string, items: ImportItem[]): Promise<ImportApplyResponse> {
  const response = await fetch('/api/import/observations/apply', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
    },
    credentials: 'include',
    body: JSON.stringify({ items }),
  })
  return parseJson<ImportApplyResponse>(response)
}
